allennlp/semparse/worlds/world.py

```diff
@@ -0,0 +1,519 @@
+        # that out right now.
+        curried_functions = self._get_curried_functions()
+        expression_type = expression.type
+        try:
+            # ``Expression.visit()`` takes two arguments: the first one is a function applied on
+            # each sub-expression and the second is a combinator that is applied to the list of
+            # values returned from the function applications. We just want the list of all
+            # sub-expressions here.
+            sub_expressions = expression.visit(lambda x: x, lambda x: x)
+            transformed_types = [sub_exp.type for sub_exp in sub_expressions]
+
+            if isinstance(expression, LambdaExpression):
+                # If the expression is a lambda expression, the list of sub expressions does not
+                # include the "lambda x" term. We're adding it here so that we will see transitions
+                # like
+                #   <e,d> -> [\x, d] instead of
+                #   <e,d> -> [d]
+                transformed_types = ["lambda x"] + transformed_types
+            elif isinstance(expression, ApplicationExpression):
+                function, arguments = expression.uncurry()
+                function_type = function.type
+                if function_type in curried_functions:
+                    expected_num_arguments = curried_functions[function_type]
+                    if len(arguments) == expected_num_arguments:
+                        # This is the initial application of a curried function.  We'll use this
+                        # node in the expression to generate the action for this function, using
+                        # all of its arguments.
+                        transformed_types = [function.type] + [argument.type for argument in arguments]
+                    else:
+                        # We're at an intermediate node.  We'll set `transformed_types` to `None`
+                        # to indicate that we need to squelch this action.
+                        transformed_types = None
+
+            if transformed_types:
+                transition = f"{expression_type} -> {transformed_types}"
+                current_transitions.append(transition)
+            for sub_expression in sub_expressions:
+                self._get_transitions(sub_expression, current_transitions)
+        except NotImplementedError:
+            # This means that the expression is a leaf. We simply make a transition from its type to itself.
+            original_name = str(expression)
+            if original_name in self.reverse_name_mapping:
+                original_name = self.reverse_name_mapping[original_name]
+            transition = f"{expression_type} -> {original_name}"
+            current_transitions.append(transition)
+        return current_transitions
+
+    def __eq__(self, other):
+        if isinstance(self, other.__class__):
+            return self.__dict__ == other.__dict__
+        return NotImplemented
```

Comment:
should this not `raise` a `NotImplementedError` rather than return one?

Reply:
Huh, I never paid attention to this.  I'm pretty sure I copied this bit from an answer on stack overflow, and didn't pay close enough attention to the return statement.  It turns out `NotImplemented` is actually a built-in object in python, so pylint didn't even complain about this.  Who knew?

I got myself into a bit of a rabbit hole figuring out exactly what this does, and why it's implemented this way.  Turns out this is the correct way to do things: https://stackoverflow.com/a/879005/1467533.